### PR TITLE
Auto reverse

### PR DESCRIPTION
**Pull request by etiennecallies**
Deployed on https://linearon.modusponens.dev/

### CODE REVIEW

File: proof.ml [hunk 1]
```diff
@@ -209,9 +208,50 @@ let from_sequent_and_rule_request_and_premises sequent rule_request premises =
...
+        with NotApplicable ->
+    try try_rule_request sequent (Promotion (get_formula_position is_ofcourse sequent))
+        with NotApplicable ->
+    try try_rule_request sequent One
+        with NotApplicable ->
+    try if List.length sequent = 1 then try_rule_request sequent (Tensor 0) else raise NotApplicable
+        with NotApplicable ->
+    try try_rule_request sequent Axiom
+        with NotApplicable ->
+    proof;;
```
Comment (etiennecallies) [inline proof.ml:246]: @olaure01 I rewrote in the same order the auto-reversible rules this way.

**Comment by etiennecallies** [comment]
Integrates #51 even if it doesn't reuse code.

File: proof.ml (hunk 1 above)
Comment (olaure01) [reply proof.ml:246]: Looks good, I am doing tests.
I would like to add one more case: `?A` with context with `Zero`, `Dual` and `Litt` only, maps to `A, ?A` with same context (i.e. apply contraction and dereliction).
I was doing it with the old version, this one is much more readable, how simple is it to do it here?

**Comment by olaure01** [comment]
Could be natural to reverse all current open premises when mode is activated. Otherwise one need to click at least once to launch it, for example on the starting sequent if one wants to start reversing from the beginning.

**Comment by olaure01** [comment]
No red turnstile on intermediary sequents. Are checks applied?
Or they could be inferred downwards a posteriori: a reversible rule with a non provable premise has a non provable conclusion.

**Comment by etiennecallies** [comment]
> Could be natural to reverse all current open premises when mode is activated. Otherwise one need to click at least once to launch it, for example on the starting sequent if one wants to start reversing from the beginning.

I agree, I'll do it on another PR.

**Comment by etiennecallies** [comment]
> No red turnstile on intermediary sequents. Are checks applied?

I agree the checks should be called on intermediary sequent. I'll do that on another PR.

> Or they could be inferred downwards a posteriori: a reversible rule with a non provable premise has a non provable conclusion.
Nice trick :) I'll see how I do that.

**Pull request merged**